### PR TITLE
Feature: add project task selection and project-based task filtering

### PR DESCRIPTION
**Pull request by IsmailMoudden**
# Feature: Add Project Selection and Project-Based Task Filtering

## Context
Previously, EurekApp allowed users to create tasks without linking them to a project.  
This led to ambiguity, weak organization, and limited collaboration in multi-project workspaces.

This PR introduces mandatory project selection and project-based task filtering, reinforcing the app’s core purpose: enabling teams to create and manage tasks within a shared project space, accessible to all members and easily filterable by project.

---

## What’s New
- Added `ProjectSelectionField` component with validation  
- Extended `CreateTaskViewModel` and `EditTaskViewModel` with project selection logic  
- Integrated `BaseTaskViewModel` for shared business rules  
- Updated navigation to include `projectId`  
- Added project filtering in `TasksScreen`  
- Added comprehensive UI and unit tests

---

## User Impact
- All tasks must now belong to a project  
- Tasks can be filtered by project  
- Clearer structure and better team collaboration

---

## Implementation Choices
- Radio buttons for simple selection with small project lists  
- Save-time validation to avoid interruptions  
- Reusable component for consistent UX

---

## Future Considerations
- Improve empty state (guide users to create their first project)  
- Cache project lists to improve performance  
- Consider searchable dropdown for 20+ projects

---

## Migration Note
Existing tasks must be migrated to include a valid `projectId`.  
Backend migration should be coordinated before deployment.

---

## Technical Improvements
- Centralize project fetching logic  
- Add tests for permission edge cases  
- Consider optimistic UI updates for better responsiveness

---

## Potential Risks
- **UI Scalability**: Radio buttons may not scale for large project lists.  
  *Mitigation*: Replace with a searchable dropdown if needed.  

- **Data Migration**: Orphaned tasks if migration fails.  
  *Mitigation*: Coordinate backend migration before merge.  

- **Permissions**: Potential edge cases during editing.  
  *Mitigation*: Add tests and refresh logic.

---

## Impact on the Product
- Stronger and more structured project–task relationship  
- Clearer navigation and better visibility across teams  
- A solid foundation for future collaborative and productivity features

---

*AI usage: AI was used to assist with parts of the implementation and to help draft the PR description. *

### CODE REVIEW

File: app/src/main/java/ch/eureka/eurekapp/screens/subscreens/tasks/CommonComponents.kt [hunk 1]
```diff
@@ -143,3 +153,51 @@ fun AttachmentsList(
...
 }
+
+@Composable
+fun ProjectSelectionField(
+    projects: List<Project>,
+    selectedProjectId: String,
+    onProjectSelected: (String) -> Unit,
+    modifier: Modifier = Modifier
+) {
+  Column(modifier = modifier) {
```
Comment (ahlombardini) [inline app/src/main/java/ch/eureka/eurekapp/screens/subscreens/tasks/CommonComponents.kt]: I would use a LazyColumn in this case. If a user belongs to many projects, this could cause significant overhead.

File: app/src/main/java/ch/eureka/eurekapp/screens/subscreens/tasks/editing/EditTaskScreen.kt
Comment (ahlombardini) [inline app/src/main/java/ch/eureka/eurekapp/screens/subscreens/tasks/editing/EditTaskScreen.kt]: Screenshot 2025-10-29 at 17 56 27
I would simply pass a task to this screen! It makes things simpler i.e.
```kotlin
fun EditTaskScreen(
    task: Task,
    navigationController: NavHostController = rememberNavController(),
    editTaskViewModel: EditTaskViewModel = viewModel(),
) {
```

File: app/src/androidTest/java/ch/eureka/eurekapp/screen/EditTaskScreenTest.kt [hunk 2]
```diff
@@ -306,7 +312,14 @@ open class EditTaskScreenTest : TestCase() {
         setupTestProject(projectId)
         setupTestTask(projectId, taskId)
 
-        val viewModel = EditTaskViewModel(taskRepository, fileRepository = FakeFileRepository())
+        val mockProjectRepository = FakeProjectRepository()
+        mockProjectRepository.setCurrentUserProjects(
+            flowOf(listOf(Project(projectId = projectId, name = "Test Project"))))
+        val viewModel =
```
Comment (IliasSokolisGD) [inline app/src/androidTest/java/ch/eureka/eurekapp/screen/EditTaskScreenTest.kt]: The idea of using mocks here is good because it enables isolation of tests

File: app/src/main/java/ch/eureka/eurekapp/model/tasks/CreateTaskState.kt [hunk 3]
```diff
@@ -14,6 +15,8 @@ data class CreateTaskState(
     override val description: String = "",
     override val dueDate: String = "",
     override val projectId: String = "",
+    override val selectedProjectId: String = "",
+    override val availableProjects: List<Project> = emptyList(),
```
[the remark this answers is not in the document]
Comment (IliasSokolisGD) [reply app/src/main/java/ch/eureka/eurekapp/model/tasks/CreateTaskState.kt]: Completely agree

File: app/src/androidTest/java/ch/eureka/eurekapp/screen/EditTaskScreenTest.kt
Comment (ahlombardini) [inline app/src/androidTest/java/ch/eureka/eurekapp/screen/EditTaskScreenTest.kt]: LGTM

File: app/src/main/java/ch/eureka/eurekapp/model/tasks/CreateTaskViewModel.kt
Comment (ahlombardini) [inline app/src/main/java/ch/eureka/eurekapp/model/tasks/CreateTaskViewModel.kt]: Great!

File: app/src/main/java/ch/eureka/eurekapp/screens/subscreens/tasks/editing/EditTaskScreen.kt
Comment (ahlombardini) [reply app/src/main/java/ch/eureka/eurekapp/screens/subscreens/tasks/editing/EditTaskScreen.kt]: In retrospect to this comment, this is fine for now, should be adressed in the future.

**Review by ahlombardini — APPROVED** [review]
Code LGTM, CI fails bcs of format, will review new push.
